cpp/src/strings/split/split.cu

```diff
@@ -385,6 +386,348 @@ struct whitespace_rsplit_tokenizer_fn
+    auto offset_buf_ptr = reinterpret_cast<size_type*>(memory_ptr);
+
+    auto d_str = d_strings.element<string_view>(idx);
+    size_type token_idx = 0;
+    size_type char_bytes_copied = 0;
+    size_type start_pos = 0;  // updates only if forward is true
+    auto end_pos = d_str.length();  // updates only if forward is false
+    while (token_idx < token_count - 1) {
+      auto delimiter_pos =
+        forward ? d_str.find(d_delimiter, start_pos) :
+                  d_str.rfind(d_delimiter, start_pos, end_pos);
+      if (delimiter_pos != -1) {
+        auto src_byte_offset =
+          forward ? d_str.byte_offset(start_pos) :
+                    d_str.byte_offset(delimiter_pos + d_delimiter.length());
+        auto token_char_bytes =
+          forward ?
+            d_str.byte_offset(delimiter_pos) - src_byte_offset :
+            d_str.byte_offset(end_pos) - src_byte_offset;
+        if (forward) {
+          for (size_type i = 0; i < token_char_bytes; ++i) {
+            char_buf_ptr[char_bytes_copied + i] =
+              d_str.data()[src_byte_offset + i];
+          }
+          offset_buf_ptr[token_idx] = char_bytes_copied;
+        }
+        else {
+          auto char_buf_offset =
+            token_size_sum - char_bytes_copied - token_char_bytes;
+          for (size_type i = 0; i < token_char_bytes; ++i) {
+            char_buf_ptr[char_buf_offset + i] =
+              d_str.data()[src_byte_offset + i];
+          }
+          offset_buf_ptr[token_count - 1 - token_idx] = char_buf_offset;
+        }
+        token_idx++;
+        char_bytes_copied += token_char_bytes;
+        if (forward) {
+          start_pos = delimiter_pos + d_delimiter.length();
+        }
+        else {
+          end_pos = delimiter_pos;
+        }
+      }
+      else {
+        break;
+      }
+    }
+    auto src_byte_offset = forward ? d_str.byte_offset(start_pos) : 0;
+    auto token_char_bytes =
+      forward ? d_str.byte_offset(end_pos) - src_byte_offset :
+                d_str.byte_offset(end_pos);
+    if (forward) {
+      for (size_type i = 0; i < token_char_bytes; ++i) {
+        char_buf_ptr[char_bytes_copied + i] = d_str.data()[src_byte_offset + i];
+      }
+      offset_buf_ptr[token_idx] = char_bytes_copied;
+    }
+    else {
+      for (size_type i = 0; i < token_char_bytes; ++i) {
+        char_buf_ptr[i] = d_str.data()[i];
+      }
+      offset_buf_ptr[0] = 0;
+    }
+    offset_buf_ptr[token_count] = token_size_sum;
+  }
+};
+
+/**
+ * @brief Compute the number of tokens, the total byte sizes of the tokens, and
+ * required memory size for the `idx'th` string element of `d_strings`.
+ */
+template <bool forward>
+struct whitespace_token_reader_fn {
+  column_device_view const d_strings;  // strings to split
+  size_type const max_tokens = std::numeric_limits<size_type>::max();
+  bool const has_validity = false;
+
+  __device__ thrust::tuple<size_type, size_type, size_type>
+  operator()(size_type idx) const {
```

Comment:
Consider refactoring `operator()` into separate calls to several `__device__` member functions for readability. Perhaps the loop might be a good candidate to factor out.